job-runner/chain/chain.go

```diff
@@ -97,103 +135,197 @@ func (c *chain) NextJobs(jobId string) proto.Jobs {
+// A chain is done running if there are no more jobs in it that can run or
+// are running. This can happen if all of the jobs in the chain are complete,
+// or if some or all of the jobs in the chain failed. Note that one failed job
+// does not mean the chain is done - there may still be pending jobs independent
+// of this failed job that can be run. Stopped jobs are treated the same as pending
+// jobs - they can be rerun (as they are when a suspended chain is resumed).
 //
 // A chain is complete if every job in it completed successfully.
-func (c *chain) IsDone() (done bool, complete bool) {
+func (c *chain) IsDoneRunning() (done bool, complete bool) {
 	c.jcMux.RLock()
 	defer c.jcMux.RUnlock()
 
 	done = true
 	complete = true
 	pendingJobs := proto.Jobs{}
 
+	// If any jobs are still running, the chain isn't done or complete.
+	for _, job := range c.jobChain.Jobs {
+		if job.State == proto.STATE_RUNNING {
+			return false, false
+		}
+	}
+
 	// Loop through every job in the chain and act on its state. Keep
 	// track of the jobs that aren't running or in a finished state so
 	// that we can later check to see if they are capable of running.
-LOOP:
 	for _, job := range c.jobChain.Jobs {
 		switch job.State {
 		case proto.STATE_RUNNING:
-			// If any jobs are running, the chain can't be done
-			// or complete, so return false for both now.
+			// We should never reach this case because we previously check for running jobs.
 			return false, false
 		case proto.STATE_COMPLETE:
 			// Move on to the next job.
-			continue LOOP
-		case proto.STATE_FAIL:
-			// This failed job is part of a sequence that can be retried
-			if c.CanRetrySequence(job) {
+			continue
+		case proto.STATE_PENDING, proto.STATE_STOPPED:
+			// Treat stopped jobs as pending jobs because they may be retried
+			// (as in a resumed job chain)
+			pendingJobs = append(pendingJobs, job)
+		default:
+			// Any job that matches none of the above cases is failed
+			if c.canRetrySequence(job.Id) {
+				// This failed job is part of a sequence that can be retried.
 				return false, false
 			}
-		default:
-			// Any job that's not running, complete, or failed.
-			pendingJobs = append(pendingJobs, job)
 		}
 
-		// We can only arrive here if a job is not complete. If there
-		// is at least one job that is not complete, the whole chain is
-		// not complete. The chain could still be done, though, so we
-		// aren't ready to return yet.
+		// We can only arrive here if a job is not complete (pending or failed).
+		// If there is at least one job that is not complete, the whole chain
+		// is not complete. The chain could still be done, though, so we aren't
+		// ready to return yet.
 		complete = false
 	}
 
-	// For each pending job, check to see if all of its previous jobs
-	// completed. If they did, there's no reason the pending job can't run.
+	// For each pending job, check to see if it's ready to be run.
 	for _, job := range pendingJobs {
```

Comment:
I think this loop is superfluous. Instead, do the check in the switch statement, no need to build a list:
```
        case proto.STATE_PENDING, proto.STATE_STOPPED:
            // Treat stopped jobs as pending jobs because they may be retried
            // (as in a resumed job chain)
            !if c.jobIsReady(job.Id) {
                return false, false
            }
```